watch/watch.go

```diff
@@ -95,6 +95,30 @@ func (idx WaitIndexVal) Next(previous BlockingParamVal) BlockingParamVal {
 	return idx
 }
 
+type WaitIndexAndLtimeVal struct {
+	Index uint64
+	Ltime uint64
+}
+
+func (wil WaitIndexAndLtimeVal) Equal(other BlockingParamVal) bool {
+	if otherWil, ok := other.(WaitIndexAndLtimeVal); ok {
+		return wil == otherWil
+	}
+	return false
+}
+
+func (wil WaitIndexAndLtimeVal) Next(previous BlockingParamVal) BlockingParamVal {
+	if previous == nil {
+		return wil
+	}
+	prevWil, ok := previous.(WaitIndexAndLtimeVal)
+	if ok && prevWil.Index > wil.Index && prevWil.Ltime > wil.Ltime {
```

Comment:
If index is effectively random then surely `prevWil.Index > wil.Index` will be true half the time. But may or may not be true in the event that Ltime actually does go backwards (e.g. agent restart).

I suspect we'd need to remove that second term but still do the reset if Ltime ever goes backwards.

I have other questions but if we decide to go ahead with this approach just wanted to note this.

Reply:
That's exactly right!

By the way, I do not understand correctly, but I would like you to tell me that the existing code will be reset when the index goes in the opposite direction, but what kind of situation is it intended for?

Whether or not index goes backwards is random.
If index does not need to be reset even if it is backwards, it may be the easiest way to prevent handler from being executed twice.

Reply:
In normal blocking the index can go backwards in some regular cases (KV listing where keys are deleted after the GC window) and generally if the servers are restored from a snapshot. It's rare but a case that needs to be handled so clients don't just get stuck in those cases.

Reply:
@banks 

Since snapshot does not hold data of event, even when server is restored from snapshot, data of event held by server will not be rewound. Therefore, it seems unnecessary to initialize the event loop.

It was impossible to reproduce the procedure where the index always goes backward in the local environment.
If there is a case where it is possible to reproduce that index surely goes backwards, it would be nice if you could tell me

I checked the [pull request](https://github.com/hashicorp/consul/pull/2621) when the index was reset.
The problem mentioned in this pull request is that when the server is restarted the client will not reconnect until a new event occurs.
For this problem, https: // github.com/hashicorp/consul/blob/master/watch/plan.go#L78 is always true for event watch.
Therefore, when an error occurs, the event loop is reset immediately, and the client is guaranteed to reconnect.

----

Although I mentioned a few times, index is not guaranteed monotonously because it is completely randomly generated by uuid.
It is shown below.

```go
// test.go
package main

import(
	"fmt"
	"strconv"

	"github.com/hashicorp/go-uuid"
)

func main(){
	for i := 1; i <= 10; i++ {
		uuid, _ := uuid.GenerateUUID()
	    fmt.Println(IDToIndex(uuid))
	}
}

// Implemented by referring to https://github.com/hashicorp/consul/blob/master/api/event.go#L92-L104
func IDToIndex(uuid string) uint64 {
	lower := uuid[0:8] + uuid[9:13] + uuid[14:18]
	upper := uuid[19:23] + uuid[24:36]
	lowVal, err := strconv.ParseUint(lower, 16, 64)
	if err != nil {
		panic("Failed to convert " + lower)
	}
	highVal, err := strconv.ParseUint(upper, 16, 64)
	if err != nil {
		panic("Failed to convert " + upper)
	}
	return lowVal ^ highVal
}
```

```
$ go run test.go
14805484469619311927
18361606342644028951
7641949632114442332
5158521267128917138
3464233816796319261
1658475967534741711
13618032119971498876
8217686710200730108
1010599251150574682
1968815759883026800
```

Therefore, whether an event loop is initialized or not is completely random.
In addition, event loops are unintentionally initialized with high probability.
Probably this is a bug.

Since the index value is random, since index goes backward, it can not be inferred what kind of action occurred.
Therefore, it is better to compare only monotonically increasing ltime and use it for index initialization. What do you think